File: tools/show_post.py
#!/usr/bin/env python3
"""Show one Bluesky post as a card in the chat, in the mode the moment calls for.

Takes an `[N]` index the model has already been shown by search-posts,
author-posts, thread-posts or liked-posts, resolves it back to an at:// URI and
emits a render payload the UI turns into a live post card.

The index is the input, not the output. It's the one thing a weak model can
reproduce reliably — it is a few digits, it was on screen a moment ago, and
getting it wrong fails loudly here instead of silently rendering someone else's
post. Nothing about the URI ever passes through the model.

**One post per call, deliberately.** Told to show two posts, the model passes
both and writes about both underneath, which stacks the cards and separates
every note from the post it is about. Taking a single ref makes the interleaving
structural instead of a rule: card, comment, card, comment. Same argument as the
tool existing at all — see PLAN.md workstream 6.

**The mode carries the intention.** A post shown for its own sake and a post
shown because the creator ought to answer it are different things on screen, and
that difference used to live only in the prose underneath ("you could reply
something like…") — where it is a suggestion nobody can act on. So the call says
which:

- `default` — the post, with what you can do with it: reply, repost, like.
- `reply` — the post with a reply box, prefilled from `--text` when the agent
  has a reply to propose. The creator edits it and sends it themselves.
- `repost` — the post with a repost box, prefilled the same way. Sending it
  empty reposts plainly; sending it written-in quotes the post.

A proposed reply is a post like any other, so `--text` is counted against the
same 300-grapheme limit a draft is (see graphemes.py) and the count travels with
the payload.

    python3 show_post.py 3
    python3 show_post.py 3 --mode reply --text "Vi gjorde samma sak i verkstan…"
"""
import argparse

import graphemes
import render
from post_index import resolve

# What the model is told once the card is drawn. A standing rule in AGENTS.md is
# thousands of tokens back by the time it matters; this arrives in the moment it
# is deciding what to do next, which is the only moment it can act on it. Both
# halves are here — say something new, and say it *now* — because both failed in
# observed runs with the rule in the instructions alone.
SEEN = ("The reader now sees the post itself — its author, its date, its full "
        "text and its pictures — so none of that needs saying again.")

NEXT = ("Write what you have to add about it now, before showing another post "
        "or doing anything else, to keep the discussion chronological. What you "
        "write next appears directly under this card; anything saved for the end "
        "of the turn arrives detached from the post it is about.")

MODES = {
    "default": SEEN + "\n" + NEXT,
    "reply": (
        SEEN + "\nUnder it is a reply box already holding the reply you passed, "
        "which the creator can edit and send. So don't write that reply out again "
        "in your own words — it is on screen. Say why this reply, and why now.\n"
        + NEXT
    ),
    "repost": (
        SEEN + "\nUnder it is a repost box holding the comment you passed, which "
        "the creator can edit, clear to repost the post plainly, or send as a "
        "quote. So don't write the comment out again — say why this post is worth "
        "putting in front of their audience.\n" + NEXT
    ),
}


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("ref", type=int, help="the post index to show, e.g. 3")
    parser.add_argument("--mode", choices=sorted(MODES), default="default",
                        help="what the card is for: default, reply or repost")
    parser.add_argument("--text", default="",
                        help="prefill for the reply or quote box")
    args = parser.parse_args()

    uri = resolve([args.ref]).get(args.ref)

    # An index that resolves to nothing is the model inventing one, and it has
    # to hear about it — a silently missing card reads as a rendering failure
    # and it will try again with the same wrong number.
    if not uri:
        print(
            f"Not shown: [{args.ref}] — no post has been listed under that index "
            "here. Only an index a search or feed tool printed in this workspace "
            "can be shown; find the post again and use the index from that output."
        )
        return

    text = args.text.strip()
    lines = [f"Showing [{args.ref}] as a card in the chat: {uri}"]

    if args.mode == "default":
        if text:
            # Said out loud rather than dropped quietly: text with no box to put
            # it in means the model meant to propose something and the creator
            # is never going to see it.
            lines.append(
                "  Ignored the text you passed — a default card has no box to put "
                "it in. Call again with mode 'reply' or 'repost' to propose it."
            )
            text = ""
    else:
        box = "reply box" if args.mode == "reply" else "quote box"
        if not text:
            lines.append(f"  An empty {box}, for the creator to write in themselves.")
        else:
            count = graphemes.count(text)
            lines.append(
                f"  {box.capitalize()} prefilled: {count}/{graphemes.LIMIT} graphemes"
                + (f" — {count - graphemes.LIMIT} OVER THE LIMIT, this cannot be "
                   "sent as written. Shorten it and show the post again."
                   if count > graphemes.LIMIT else "")
            )

    print("\n".join(lines))
    print("\n" + MODES[args.mode])

    render.emit(
        "posts",
        posts=[{"ref": args.ref, "uri": uri}],
        mode=args.mode,
        text=text,
        limit=graphemes.LIMIT,
    )


if __name__ == "__main__":
    main()

File: agent/.opencode/tools/show-post.ts
import { tool } from "@opencode-ai/plugin"
import { $ } from "bun"

// Thin wrapper around tools/show_post.py — located via the git worktree so
// there are no hardcoded paths.
//
// The one tool here that exists for the *interface* rather than for data: it
// fetches nothing and calls no API. It's how the model asks the UI to draw
// something, through the channel it is most reliable in.
//
// Takes ONE ref, not a list. Given a list the model shows every post at once
// and writes about them all underneath, so each note ends up nowhere near its
// post. A single argument makes card-then-comment the only shape available.
export default tool({
  description:
    "Show one Bluesky post as a live card in the chat. Pass the `[N]` index " +
    "you were given by search-posts, author-posts, thread-posts or liked-posts " +
    "— e.g. ref: 3. The card appears where you made the call and whatever you " +
    "write next appears under it. Showing several posts means calling this " +
    "several times, each call followed by your comment on that one post. " +
    "Use `mode` to say what the post is being shown for: 'reply' or 'repost' " +
    "turn the card into something the creator can act on, with your proposed " +
    "words already in the box. " +
    "This is the only way to show a post. Two things never go in your reply: " +
    "the index (the reader has never seen it) and anything already on the card " +
    "— the author, the date, the counts, the post's own text and any reply you " +
    "passed in `text` are all right there, so write only what you have to add.",
  args: {
    ref: tool.schema.number().describe("The index of the post to show, e.g. 3"),
    mode: tool.schema
      .enum(["default", "reply", "repost"])
      .describe(
        "What the card is for. 'default': the post with reply, repost and like " +
          "buttons. 'reply': the post with a reply box the creator edits and " +
          "sends — put the reply you propose in `text`. 'repost': the post with " +
          "a repost box, which reposts plainly if left empty and quotes the post " +
          "if written in — put the comment you propose in `text`. " +
          "Use reply or repost whenever you are " +
          "suggesting the creator engage; a suggestion written in prose is one " +
          "they cannot act on.",
      )
      .optional(),
    text: tool.schema
      .string()
      .describe(
        "The reply or quote comment you are proposing, in the creator's own " +
          "language and voice. It is the actual message, not a description of " +
          "one, and it is theirs to edit before sending. Only used with mode " +
          "'reply' or 'repost'.",
      )
      .optional(),
  },
  async execute(args, context) {
    const script = `${context.worktree}/tools/show_post.py`
    // The model sometimes sends "3" or "[3]" rather than 3, and argparse's int
    // type would exit 2 on either. Coerce here so a stray bracket costs
    // nothing. An array slips through occasionally too — take the first.
    const raw = Array.isArray(args.ref) ? args.ref[0] : args.ref
    const ref = parseInt(String(raw).replace(/[^\d-]/g, ""), 10)
    if (!Number.isFinite(ref)) return "No post index given — pass ref like 3."
    // An unknown mode is a typo, not a reason to fail the call: the card is
    // still worth drawing, and default is the mode that shows everything.
    const known = ["default", "reply", "repost"]
    const mode = known.includes(String(args.mode)) ? String(args.mode) : "default"
    const text = args.text ? ["--text", String(args.text)] : []
    return await $`python3 ${script} ${String(ref)} --mode ${mode} ${text}`.env({ ...process.env, ACTING_SESSION: context.sessionID }).text()
  },
})
